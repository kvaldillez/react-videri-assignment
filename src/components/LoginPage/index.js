import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import bgImage from '../../assets/images/bg-login.png';

const Container = styled.div`
  height: 100vh;
  background: url(${bgImage}) no-repeat center;
  background-size: cover;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Content = styled.div`
  padding: 5rem 10rem;
  background-color: #fff;
  flex: 0 0 auto;
  width: 60%;
  max-width: 65rem;

  @media (max-width: 900px) {
    padding: 5rem;
  }

  @media (max-width: 600px) {
    padding: 3rem;
    width: 80%;
  }
`;

const Title = styled.h1`
  margin: 0 0 3rem;
  text-transform: uppercase;

  span {
    font-size: 3.2rem;
    letter-spacing: 1rem;
  }
`;

const Subtitle = styled.h2`
  margin: 0 0 2rem;
  text-transform: uppercase;
  font-weight: bold;
  color: #000;
`;

const Form = styled.form`
  display: flex;
  flex-direction: column;
`;

const Input = styled.input`
  margin: 0 0 2rem;
  padding: 0 1.5rem;
  height: 4.5rem;
  border: 0.2rem solid #ddd;
  font-size: 1.4rem;

  ::placeholder {
    font-weight: bold;
    text-transform: uppercase;
    opacity: 0.5;
  }

  &.valid {
    border-color: #34c064;
  }

  &.invalid {
    border-color: red;
  }
`;

const Button = styled.button`
  margin: 0 0 2rem;
  padding: 1.2rem;
  width: 19.6rem;
  align-self: flex-end;
  font-size: 1.4rem;
  font-weight: bold;
  background-color: #34c064;
  text-transform: uppercase;
  cursor: pointer;
  color: #fff;
`;

const Error = styled.p`
  margin: 0 0 2rem;
  color: red;
`;

const LoginPage = props => (
  <Container>
    <Content>
      <Title>
        <span>Videri</span> Orchestrator
      </Title>
      <Subtitle>Sign In</Subtitle>
      <Form onSubmit={props.handleSubmitForm}>
        {props.formControlsArray.map(formElement => (
          <Input
            key={formElement.id}
            type={formElement.config.type}
            className={
              formElement.config.touched &&
              formElement.config.validation &&
              formElement.config.valid
                ? 'valid'
                : formElement.config.touched
                  ? 'invalid'
                  : null
            }
            placeholder={formElement.config.placeholder}
            onChange={e => props.handleChangeInput(e, formElement.id)}
          />
        ))}
        {props.displayErrorMessage && <Error>{props.errorMessage}</Error>}
        <Button type="submit">Sign In</Button>
      </Form>
    </Content>
  </Container>
);

LoginPage.propTypes = {
  formControlsArray: PropTypes.arrayOf(
    PropTypes.shape({
      src: PropTypes.string,
      alt: PropTypes.string,
      isActive: PropTypes.bool,
    })
  ).isRequired,
  handleSubmitForm: PropTypes.func.isRequired,
  handleChangeInput: PropTypes.func.isRequired,
  displayErrorMessage: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
};

export default LoginPage;
