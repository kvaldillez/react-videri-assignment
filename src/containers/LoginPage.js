import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import LoginPageComponent from '../components/LoginPage';
import { checkValidity } from '../utilities';
import { loginUser } from '../store/actions';

export class LoginPageContainer extends Component {
  static propTypes = {
    loading: PropTypes.bool.isRequired,
  };

  state = {
    formControls: {
      username: {
        type: 'email',
        placeholder: 'USERNAME',
        value: '',
        validation: {
          required: true,
          isEmail: true,
        },
        valid: false,
        touched: false,
      },
      password: {
        type: 'password',
        placeholder: 'PASSWORD',
        value: '',
        validation: {
          required: true,
          isPassword: true,
          minLength: 8,
        },
        valid: false,
        touched: false,
      },
    },
    displayErrorMessage: false,
    errorMessage: '',
  };

  componentDidMount() {}

  handleChangeInput = (e, key) => {
    const updatedControls = {
      ...this.state.formControls,
      [key]: {
        ...this.state.formControls[key],
        value: e.target.value,
        touched: true,
        valid: checkValidity(
          e.target.value,
          this.state.formControls[key].validation
        ),
      },
    };

    this.setState({ formControls: updatedControls });
  };

  handleSubmitForm = e => {
    e.preventDefault();

    let submitForm = true;
    const errors = [];

    for (const key in this.state.formControls) {
      if (!this.state.formControls[key].valid) {
        errors.push(`${key}`);
        submitForm = false;
      }
    }

    this.setState({
      errorMessage: `Please fix the ${errors.join(' and ')} input(s) above.`,
    });

    if (submitForm) {
      this.setState({ displayErrorMessage: false });
      this.props.loginUser(this.state.formControls.username.value);
    } else {
      this.setState({ displayErrorMessage: true });
    }
  };

  render() {
    const formControlsArray = [];

    for (const key in this.state.formControls) {
      formControlsArray.push({
        id: key,
        config: this.state.formControls[key],
      });
    }

    return (
      <LoginPageComponent
        loading={this.props.loading}
        formControlsArray={formControlsArray}
        displayErrorMessage={this.state.displayErrorMessage}
        errorMessage={this.state.errorMessage}
        handleSubmitForm={this.handleSubmitForm}
        handleChangeInput={this.handleChangeInput}
      />
    );
  }
}

const mapStateToProps = state => ({
  loading: state.loading,
});

const mapDispatchToProps = dispatch => ({
  loginUser: id => dispatch(loginUser.request(id)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginPageContainer);
