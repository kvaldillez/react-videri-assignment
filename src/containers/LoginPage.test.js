import { LoginPageContainer } from './LoginPage';

describe('LoginPage', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(<LoginPageContainer loading={false} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('has all input fields', () => {
    const wrapper = mount(<LoginPageContainer loading={false} />);
    let fields = wrapper.find('form').find('input').length;
    expect(fields).toEqual(2);
  });

  it('has a submit form button', () => {
    const wrapper = mount(<LoginPageContainer loading={false} />);
    let button = wrapper.find('form').find('button[type="submit"]').length;
    expect(button).toEqual(1);
  });
});
