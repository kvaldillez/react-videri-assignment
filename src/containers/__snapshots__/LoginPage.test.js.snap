// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage renders without crashing 1`] = `
<LoginPage
  displayErrorMessage={false}
  errorMessage=""
  formControlsArray={
    Array [
      Object {
        "config": Object {
          "placeholder": "USERNAME",
          "touched": false,
          "type": "email",
          "valid": false,
          "validation": Object {
            "isEmail": true,
            "required": true,
          },
          "value": "",
        },
        "id": "username",
      },
      Object {
        "config": Object {
          "placeholder": "PASSWORD",
          "touched": false,
          "type": "password",
          "valid": false,
          "validation": Object {
            "isPassword": true,
            "minLength": 8,
            "required": true,
          },
          "value": "",
        },
        "id": "password",
      },
    ]
  }
  handleChangeInput={[Function]}
  handleSubmitForm={[Function]}
  loading={false}
/>
`;
